### PR TITLE
invoices/sqldb: query by ChanID when updating AMP invoice preimage

### PR DESCRIPTION
**Pull request by aakselrod**
## Change Description

This PR fixes the following test:

```
make itest backend=btcd dbbackend=postgres nativesql=true icase=sendpayment_amp_invoice
```

Note that the `sendpayment_amp_invoice_repeat` test is not fixed by this. Also note that this depends on #9021 to unmask the test failure.

## Steps to Test

Run the itest as above before and after applying this PR.

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

**Review by bhandras — APPROVED** [review]
Very nice catch! Thank you! LGTM 🎉

**Comment by bhandras** [comment]
Please make sure that the linter is green too!

File: invoices/sql_store.go [hunk 1]
```diff
@@ -1116,6 +1116,9 @@ func (s *sqlInvoiceUpdater) AddAmpHtlcPreimage(setID [32]byte,
 			SetID:     setID[:],
 			HtlcID:    int64(circuitKey.HtlcID),
 			Preimage:  preimage[:],
+			ChanID: strconv.FormatUint(
```
Comment (yyforyongyu) [inline invoices/sql_store.go:1119]: Does this mean we previously may have updated the wrong htlc?

Comment (aakselrod) [reply invoices/sql_store.go:1119]: Yes, with sqlite we updated multiple HTLCs at once with the same preimage. With postgres, we never updated any preimages at all.

**Comment by aakselrod** [comment]
> Please make sure that the linter is green too!

I have questions on this:

The lint task fails on checking submodule tidiness. However, locally, I can't reproduce this:

```
$ git log -n 1
commit a6e1960c61bd11b4ab9884e26c4814b59840df4a (HEAD -> add-channelid-ampsettle, alex/add-channelid-ampsettle)
Author: Alex Akselrod <alex@akselrod.org>
Date:   Wed Aug 21 17:33:37 2024 -0700

    invoices/sqldb: query by ChanID when updating AMP invoice preimage
$ git status
On branch add-channelid-ampsettle
nothing to commit, working tree clean
$ ./scripts/tidy_modules.sh 
~/go/src/github.com/lightningnetwork/lnd/queue ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/kvdb ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/tools ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/ticker ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/tlv ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/clock ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/healthcheck ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/fn ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/sqldb ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/cert ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd/tor ~/go/src/github.com/lightningnetwork/lnd
~/go/src/github.com/lightningnetwork/lnd
$ git status
On branch add-channelid-ampsettle
nothing to commit, working tree clean
$ 
```

The other thing, when I `make lint`, it doesn't like me doing a local redirect for `sqldb`. Would love a suggestion for how to fix that.
```
$ make lint
 Checking for target Go version (v1.22.6) in Dockerfile files (*Dockerfile)
./scripts/check-go-version-dockerfile.sh 1.22.6
./docker/btcd/Dockerfile is using Go version 1.22.6.
Skipping ./tools/Dockerfile
./make/builder.Dockerfile is using Go version 1.22.6.
./lnrpc/Dockerfile is using Go version 1.22.6.
./dev.Dockerfile is using Go version 1.22.6.
./Dockerfile is using Go version 1.22.6.
All Dockerfiles pass the Go version check for Go version 1.22.6.
 Checking for target Go version (v1.22.6) in YAML files (*.yaml, *.yml)
./scripts/check-go-version-yaml.sh 1.22.6
./.golangci.yml specifies Go version 1.22.6.
./.github/workflows/release.yaml specifies Go version 1.22.6.
./.github/workflows/main.yml specifies Go version 1.22.6.
All YAML files pass the Go version check for Go version 1.22.6.
 Building tools docker image.
docker build -q -t lnd-tools tools
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            Install the buildx component to build images with BuildKit:
            https://docs.docker.com/go/buildx/

sha256:0c7cd8c3d3aac815f14073c8a6483a4e26686231dfaae460503ceb59178db57c
 Linting source.
docker run --rm -v /home/user/.cache/go-build:/tmp/build/.cache -v /home/user/go/pkg/mod:/tmp/build/.modcache -v /tmp/go-lint-cache:/root/.cache/golangci-lint -v $(pwd):/build lnd-tools golangci-lint run -v 
level=info msg="golangci-lint has version v1.60.1 built with go1.22.6 from (unknown, modified: ?, mod sum: \"h1:DRKNqNTQRLBJZ1il5u4fvgLQCjQc7QFs0DbhksJtVJE=\") on (unknown)"
level=info msg="[config_reader] Config search paths: [./ /build / /root]"
level=info msg="[config_reader] Used config file .golangci.yml"
level=warning msg="The linter 'execinquery' is deprecated (since v1.58.0) due to: The repository of the linter has been archived by the owner. "
level=info msg="[lintersdb] Active 65 linters: [asasalint asciicheck bidichk canonicalheader containedctx decorder dupl dupword durationcheck errcheck errchkjson errorlint execinquery exportloopref fatcontext forbidigo forcetypeassert funlen gci ginkgolinter gocheckcompilerdirectives gochecksumtype goconst gocritic godot gofmt goheader goimports gomoddirectives gomodguard goprintffuncname gosec gosimple govet grouper importas ineffassign interfacebloat lll loggercheck makezero mirror misspell musttag nakedret nestif nlreturn nolintlint nonamedreturns nosprintfhostport predeclared promlinter reassign sloglint spancheck staticcheck stylecheck tagliatelle tenv testableexamples unconvert unused usestdlibvars whitespace zerologlint]"
level=info msg="[loader] Using build tags: [autopilotrpc chainrpc dev invoicesrpc neutrinorpc peersrpc signrpc walletrpc watchtowerrpc kvdb_etcd kvdb_postgres kvdb_sqlite integration]"
level=info msg="[loader] Go packages loading at mode 575 (exports_file|files|name|types_sizes|compiled_files|deps|imports) took 741.598014ms"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 257.894938ms"
level=info msg="[linters_context] importas settings found, but no aliases listed. List aliases under alias: key."
level=info msg="[linters_context/goanalysis] analyzers took 0s with no stages"
level=warning msg="[runner/nolint] Found unknown linters in //nolint directives: ifelsechain"
level=info msg="[runner] Issues before processing: 31379, after processing: 1"
level=info msg="[runner] Processors filtering stat (in/out): path_shortener: 1/1, filename_unadjuster: 31379/31379, invalid_issue: 31379/31379, path_prettifier: 31379/31379, skip_dirs: 31379/31379, identifier_marker: 9917/9917, max_from_linter: 1/1, skip_files: 31379/31379, nolint: 8073/6980, diff: 6815/1, path_prefixer: 1/1, sort_results: 1/1, uniq_by_line: 6980/6815, max_per_file_from_linter: 1/1, max_same_issues: 1/1, source_code: 1/1, cgo: 31379/31379, autogenerated_exclude: 31379/9917, exclude: 9917/9917, exclude-rules: 9917/8073, severity-rules: 1/1, fixer: 1/1"
level=info msg="[runner] processing took 870.841685ms with stages: nolint: 419.503652ms, exclude-rules: 203.227664ms, identifier_marker: 139.133924ms, diff: 47.532949ms, autogenerated_exclude: 23.356419ms, path_prettifier: 17.32278ms, invalid_issue: 11.46177ms, skip_dirs: 5.796463ms, uniq_by_line: 1.345988ms, cgo: 1.214543ms, filename_unadjuster: 883.714µs, source_code: 56.572µs, max_same_issues: 1.545µs, path_shortener: 1.019µs, max_per_file_from_linter: 670ns, max_from_linter: 586ns, fixer: 322ns, skip_files: 306ns, exclude: 264ns, sort_results: 245ns, path_prefixer: 199ns, severity-rules: 91ns"
level=info msg="[runner] linters took 1.702902941s with stages: goanalysis_metalinter: 831.933394ms"
level=info msg="File cache stats: 1 entries of total size 10.3KiB"
go.mod:208:1: local replacement are not allowed: github.com/lightningnetwork/lnd/sqldb (gomoddirectives)
replace github.com/lightningnetwork/lnd/sqldb => ./sqldb
^
level=info msg="Memory: 29 samples, avg is 91.2MB, max is 132.9MB"
level=info msg="Execution took 2.707028088s"
make: *** [Makefile:322: lint-source] Error 1
$
```

**Comment by guggero** [comment]
@aakselrod I get the following diff when I run `make tidy-module-check`:

``` diff
diff --git a/go.mod b/go.mod
index 6652e6547..cbe4d40ba 100644
--- a/go.mod
+++ b/go.mod
@@ -19,7 +19,6 @@ require (
        github.com/davecgh/go-spew v1.1.1
        github.com/decred/dcrd/dcrec/secp256k1/v4 v4.3.0
        github.com/go-errors/errors v1.0.1
-       github.com/golang/protobuf v1.5.3
        github.com/gorilla/websocket v1.5.0
        github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
        github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
@@ -95,6 +94,7 @@ require (
        github.com/gogo/protobuf v1.3.2 // indirect
        github.com/golang-jwt/jwt/v4 v4.4.2 // indirect
        github.com/golang-migrate/migrate/v4 v4.17.0 // indirect
+       github.com/golang/protobuf v1.5.3 // indirect
        github.com/golang/snappy v0.0.4 // indirect
        github.com/google/btree v1.0.1 // indirect
        github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510 // indirect
```

Maybe a question of the Go version you're using?


```
go.mod:208:1: local replacement are not allowed: github.com/lightningnetwork/lnd/sqldb (gomoddirectives)
replace github.com/lightningnetwork/lnd/sqldb => ./sqldb
```

Hmm, normally we do this by doing a PR to the submodule first, merge that and push a tag, then bump the version in the main `go.mod` file. But I guess since that module is still under heavy development, I'd also agree to add this to the allowed replace modules in `.golangci.yml`:

```
  gomoddirectives:
    replace-local: true
    replace-allow-list:
...
```

**Review by yyforyongyu — COMMENTED** [review]
`make tidy-module-check` is different from running the script directly as it involves one more check,
https://github.com/lightningnetwork/lnd/blob/cc9e2b783ec983fbe3995d47d31898aa122b11c8/Makefile#L337-L339

File: invoices/sql_store.go (hunk 1 above)
Comment (yyforyongyu) [reply invoices/sql_store.go:1119]: I think this only affects AMP HTLCs - could you add a release note on this fix?

**Comment by aakselrod** [comment]
> `make tidy-module-check` is different from running the script directly as it involves one more check,
> 
> https://github.com/lightningnetwork/lnd/blob/cc9e2b783ec983fbe3995d47d31898aa122b11c8/Makefile#L337-L339

That did it, thanks!

**Comment by aakselrod** [comment]
> @aakselrod I get the following diff when I run `make tidy-module-check`:

I think my problem was that I wasn't running the `make` target but the script directly.

> Hmm, normally we do this by doing a PR to the submodule first, merge that and push a tag, then bump the version in the main `go.mod` file. But I guess since that module is still under heavy development, I'd also agree to add this to the allowed replace modules in `.golangci.yml`:
> 
> ```
>   gomoddirectives:
>     replace-local: true
>     replace-allow-list:
> ...
> ```

OK, I'll make this change. Thank you!

File: invoices/sql_store.go (hunk 1 above)
Comment (aakselrod) [reply invoices/sql_store.go:1119]: Yes, only AMP HTLCs with multiple HTLCs sharing a set ID. Will update release notes.

Comment (aakselrod) [reply invoices/sql_store.go:1119]: Updated release notes.

**Review by yyforyongyu — APPROVED** [review]
Thanks for the fix!

**Comment by guggero** [comment]
I think we should do a full CI run again once https://github.com/lightningnetwork/lnd/pull/9021 is in and this is rebased.

**Comment by aakselrod** [comment]
I'll do that rebase today!

**Comment by aakselrod** [comment]
Rebase complete, hopefully should be good to merge after CI run.

**Review by yyforyongyu — APPROVED** [review]
Thanks again - confirmed the CI failures are flakes, which will be fixed in 0.19.0.

**Review by guggero — CHANGES_REQUESTED** [review]
Actually, this test failure very much looks like it could be affected by these changes:
```
     --- FAIL: TestLightningNetworkDaemon/tranche05/107-of-168/bitcoind/sendpayment_amp_invoice_repeat (19.61s)
        lnd_amp_test.go:300: 
            	Error Trace:	/home/runner/work/lnd/lnd/itest/lnd_amp_test.go:300
            	            				/home/runner/work/lnd/lnd/lntest/harness.go:386
            	            				/home/runner/work/lnd/lnd/itest/lnd_test.go:139
            	Error:      	Not equal: 
            	            	expected: 1
            	            	actual  : 2
            	Test:       	TestLightningNetworkDaemon/tranche05/107-of-168/bitcoind/sendpayment_amp_invoice_repeat
```

I ran the test locally and got the following errors in the Postgres container:

```
2024-08-26 08:45:15.430 UTC [88] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
2024-08-26 08:45:15.430 UTC [88] DETAIL:  Key (payment_addr)=(\x68d295d5d6daa959147608463bb46f5ae877f0806d3af93ff48f681e79411700) already exists.
2024-08-26 08:45:15.430 UTC [88] STATEMENT:  -- name: InsertInvoice :one
        INSERT INTO invoices (
            hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
            payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
            is_hodl, is_keysend, created_at
        ) VALUES (
            $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
        ) RETURNING id
        
2024-08-26 08:45:15.440 UTC [88] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
2024-08-26 08:45:15.440 UTC [88] DETAIL:  Key (payment_addr)=(\x68d295d5d6daa959147608463bb46f5ae877f0806d3af93ff48f681e79411700) already exists.
2024-08-26 08:45:15.440 UTC [88] STATEMENT:  -- name: InsertInvoice :one
        INSERT INTO invoices (
            hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
            payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
            is_hodl, is_keysend, created_at
        ) VALUES (
            $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
        ) RETURNING id
        
2024-08-26 08:45:15.686 UTC [88] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
2024-08-26 08:45:15.686 UTC [88] DETAIL:  Key (payment_addr)=(\x68d295d5d6daa959147608463bb46f5ae877f0806d3af93ff48f681e79411700) already exists.
2024-08-26 08:45:15.686 UTC [88] STATEMENT:  -- name: InsertInvoice :one
        INSERT INTO invoices (
            hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
            payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
            is_hodl, is_keysend, created_at
        ) VALUES (
            $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
        ) RETURNING id
        
2024-08-26 08:46:10.586 UTC [346] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
2024-08-26 08:46:10.586 UTC [346] DETAIL:  Key (payment_addr)=(\xc781996e8a209bfdd5e2775876b8626da735bfb088a94423985f8307a7aae2ac) already exists.
2024-08-26 08:46:10.586 UTC [346] STATEMENT:  -- name: InsertInvoice :one
        INSERT INTO invoices (
            hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
            payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
            is_hodl, is_keysend, created_at
        ) VALUES (
            $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
        ) RETURNING id
        
2024-08-26 08:46:10.806 UTC [346] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
2024-08-26 08:46:10.806 UTC [346] DETAIL:  Key (payment_addr)=(\xc781996e8a209bfdd5e2775876b8626da735bfb088a94423985f8307a7aae2ac) already exists.
2024-08-26 08:46:10.806 UTC [346] STATEMENT:  -- name: InsertInvoice :one
        INSERT INTO invoices (
            hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
            payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
            is_hodl, is_keysend, created_at
        ) VALUES (
            $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
        ) RETURNING id
```

**Comment by aakselrod** [comment]
These errors are caused by https://github.com/lightningnetwork/lnd/blob/8939a217c3f1bb0b56a3a77df734fc6c74347838/invoices/invoiceregistry.go#L888 and ignored. 3 of them are from the test fixed by this PR. The remainder are from the next test, explicitly not fixed by this PR (see OP change description), which I'm still debugging and for which I was planning to submit a follow-up PR.

**Comment by guggero** [comment]
> explicitly not fixed by this PR (see OP change description)

Ah, right. I missed that part, sorry.

> was planning to submit a follow-up PR.

Sounds good. Though I think we'll want to block `v0.18.3` on this fix. Or do you think it might be an issue of the itest itself and not the actual native SQL code?

**Comment by aakselrod** [comment]
I think it's an issue with invoice numbering, the behavior is different between kvdb and sql here, so it looks like a real but separate bug to me.

**Comment by aakselrod** [comment]
The `invoice_exact_end_date` failure is caused by the way KVDB compares timestamps (integer seconds) vs SQL (microseconds), but not sure whether that should be fixed in the test, KVDB, or SQL, BTW. It's likely not as much of a blocker.

**Comment by Roasbeef** [comment]
Looks like the native SQL itest are still failing: 
```
  --- FAIL: TestLightningNetworkDaemon/tranche05/107-of-168/bitcoind/sendpayment_amp_invoice_repeat (19.61s)
        lnd_amp_test.go:300: 
            	Error Trace:	/home/runner/work/lnd/lnd/itest/lnd_amp_test.go:300
            	            				/home/runner/work/lnd/lnd/lntest/harness.go:386
            	            				/home/runner/work/lnd/lnd/itest/lnd_test.go:139
            	Error:      	Not equal: 
            	            	expected: 1
            	            	actual  : 2
            	Test:       	TestLightningNetworkDaemon/tranche05/107-of-168/bitcoind/sendpayment_amp_invoice_repeat
```

**Comment by aakselrod** [comment]
Didn't get too much time today to track down the `sendpayment_amp_invoice_repeat` failure, but this is why I mentioned above that I think it has to do with invoice numbering (see specifically Dave, whose invoices are the ones we're checking):

```
$ grep -R "missed an update" .*
.logs-tranche3/0-single-hop_send_to_route-Alice-0307f3a4.log:2024-08-26 08:18:33.744 [WRN] INVC: client=1 for invoice notifications missed an update, add_index=0, new add event index=7
.logs-tranche5/9-sendpayment_amp_invoice_repeat-Dave-02424a66.log:2024-08-26 08:01:17.330 [WRN] INVC: client=1 for invoice notifications missed an update, add_index=1, new add event index=1
.logs-tranche5/9-sendpayment_amp_invoice_repeat-Dave-02424a66.log:2024-08-26 08:01:17.891 [WRN] INVC: client=1 for invoice notifications missed an update, add_index=1, new add event index=1
.logs-tranche5/1-sendpayment_amp_invoice_repeat-Bob-038abfe7.log:2024-08-26 07:57:58.019 [WRN] INVC: client=1 for invoice notifications missed an update, add_index=1, new add event index=1
.logs-tranche5/1-sendpayment_amp_invoice_repeat-Bob-038abfe7.log:2024-08-26 08:00:05.123 [WRN] INVC: client=1 for invoice notifications missed an update, add_index=0, new add event index=5
.logs-tranche5/1-sendpayment_amp_invoice_repeat-Bob-038abfe7.log:2024-08-26 08:00:06.561 [WRN] INVC: client=1 for invoice notifications missed an update, add_index=6, new add event index=6
```

Hopefully will get more time tomorrow to debug.

**Comment by aakselrod** [comment]
> I think it's an issue with invoice numbering, the behavior is different between kvdb and sql here, so it looks like a real but separate bug to me.

I no longer think it's about invoice numbering but didn't have much time to keep digging into it today. Will keep going tomorrow.

**Comment by aakselrod** [comment]
I've spent a few hours on this today, and have gotten farther. It looks like the SQL invoice updater isn't returning only the "projected" HTLCs for the newly-settled AMP invoice, but all of the past invoices. The newly-settled AMP invoice is also returned with the incorrect settle index and time, though it seems to be written correctly to the DB and returned correctly on `LookupInvoice` after the fact.

I also noticed that in the KVDB code, the updater consistently logs the error

```
[ERR] CHDB: unable to fetch amp htlcs for inv [0 0 0 0] and setIDs [0xc0007ea100]: %!w(*errors.errorString=&{unable to locate invoice})
```

Not 100% sure that's relevant, but I'm going to keep going on this tomorrow.

**Comment by aakselrod** [comment]
Also will take a closer look at https://github.com/lightningnetwork/lnd/blob/1bf7ad9b435ae6a5a1aaecdab77934b59936495d/invoices/invoices_test.go#L234 and see if this runs native sql tests.

**Comment by bhandras** [comment]
> I've spent a few hours on this today, and have gotten farther. It looks like the SQL invoice updater isn't returning only the "projected" HTLCs for the newly-settled AMP invoice, but all of the past invoices. The newly-settled AMP invoice is also returned with the incorrect settle index and time, though it seems to be written correctly to the DB and returned correctly on `LookupInvoice` after the fact.
> 
> I also noticed that in the KVDB code, the updater consistently logs the error
> 
> ```
> [ERR] CHDB: unable to fetch amp htlcs for inv [0 0 0 0] and setIDs [0xc0007ea100]: %!w(*errors.errorString=&{unable to locate invoice})
> ```
> 
> Not 100% sure that's relevant, but I'm going to keep going on this tomorrow.

I believe the issue is that when we fetch the invoice during the update the `setID` passed to `UpdateInvoice` https://github.com/lightningnetwork/lnd/blob/5c3a8e949c7d30ab055fd4326a2082ad3b0f7b0f/invoices/sql_store.go#L1301 is not used. I think simply passing in the `setID` won't work as the SQL query assumes that the AMP sub-invoice already exists. In the KV version the assumption is different as if the `setID` is passed it'll simply fetch the HTLCs in that set (https://github.com/lightningnetwork/lnd/blob/5c3a8e949c7d30ab055fd4326a2082ad3b0f7b0f/channeldb/invoices.go#L1508). I can try to put together a diff to fix this.

**Comment by aakselrod** [comment]
> I believe the issue is that when we fetch the invoice during the update the `setID` passed to `UpdateInvoice` is not used:

This looks exactly right to me.

> I can try to put together a diff to fix this.

Sounds good, thank you! Do you want to make a separate PR or add it to this one?

**Comment by bhandras** [comment]
> > I believe the issue is that when we fetch the invoice during the update the `setID` passed to `UpdateInvoice` is not used:
> 
> This looks exactly right to me.
> 
> > I can try to put together a diff to fix this.
> 
> Sounds good, thank you! Do you want to make a separate PR or add it to this one?

I think it's best to add to this one, given it fails the itest.

**Comment by aakselrod** [comment]
> The newly-settled AMP invoice is also returned with the incorrect settle index and time, though it seems to be written correctly to the DB and returned correctly on LookupInvoice after the fact.

I've pushed an itest change that checks for this, and then a fix. Hopefully they both run so we can see the difference...

**Comment by guggero** [comment]
Did a quick rebase to fix a merge conflict in the release notes, so we can see the CI run.

**Comment by bhandras** [comment]
Added fixes for the remaining issues. Note that the fix changes some previous behavior IMO for the better (cc @Roasbeef)

**Comment by bhandras** [comment]
I wonder if we should instead open another PR with the rest of the fixes since these seem to be unrelated to the original fix.

**Comment by bhandras** [comment]
> I ran the test locally and got the following errors in the Postgres container:
> 
> ```
> 2024-08-26 08:45:15.430 UTC [88] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
> 2024-08-26 08:45:15.430 UTC [88] DETAIL:  Key (payment_addr)=(\x68d295d5d6daa959147608463bb46f5ae877f0806d3af93ff48f681e79411700) already exists.
> 2024-08-26 08:45:15.430 UTC [88] STATEMENT:  -- name: InsertInvoice :one
>         INSERT INTO invoices (
>             hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
>             payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
>             is_hodl, is_keysend, created_at
>         ) VALUES (
>             $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
>         ) RETURNING id
>         
> 2024-08-26 08:45:15.440 UTC [88] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
> 2024-08-26 08:45:15.440 UTC [88] DETAIL:  Key (payment_addr)=(\x68d295d5d6daa959147608463bb46f5ae877f0806d3af93ff48f681e79411700) already exists.
> 2024-08-26 08:45:15.440 UTC [88] STATEMENT:  -- name: InsertInvoice :one
>         INSERT INTO invoices (
>             hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
>             payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
>             is_hodl, is_keysend, created_at
>         ) VALUES (
>             $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
>         ) RETURNING id
>         
> 2024-08-26 08:45:15.686 UTC [88] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
> 2024-08-26 08:45:15.686 UTC [88] DETAIL:  Key (payment_addr)=(\x68d295d5d6daa959147608463bb46f5ae877f0806d3af93ff48f681e79411700) already exists.
> 2024-08-26 08:45:15.686 UTC [88] STATEMENT:  -- name: InsertInvoice :one
>         INSERT INTO invoices (
>             hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
>             payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
>             is_hodl, is_keysend, created_at
>         ) VALUES (
>             $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
>         ) RETURNING id
>         
> 2024-08-26 08:46:10.586 UTC [346] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
> 2024-08-26 08:46:10.586 UTC [346] DETAIL:  Key (payment_addr)=(\xc781996e8a209bfdd5e2775876b8626da735bfb088a94423985f8307a7aae2ac) already exists.
> 2024-08-26 08:46:10.586 UTC [346] STATEMENT:  -- name: InsertInvoice :one
>         INSERT INTO invoices (
>             hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
>             payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
>             is_hodl, is_keysend, created_at
>         ) VALUES (
>             $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
>         ) RETURNING id
>         
> 2024-08-26 08:46:10.806 UTC [346] ERROR:  duplicate key value violates unique constraint "invoices_payment_addr_key"
> 2024-08-26 08:46:10.806 UTC [346] DETAIL:  Key (payment_addr)=(\xc781996e8a209bfdd5e2775876b8626da735bfb088a94423985f8307a7aae2ac) already exists.
> 2024-08-26 08:46:10.806 UTC [346] STATEMENT:  -- name: InsertInvoice :one
>         INSERT INTO invoices (
>             hash, preimage, memo, amount_msat, cltv_delta, expiry, payment_addr, 
>             payment_request, payment_request_hash, state, amount_paid_msat, is_amp,
>             is_hodl, is_keysend, created_at
>         ) VALUES (
>             $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15
>         ) RETURNING id
> ```

Note that these logs are harmless and emitted when running `icase=send_payment_amp`. In this scenario Bob adds an AMP invoice on-demand (spontaneous AMP) and it is expected that subsequent `InsertInvoice` calls will fail. See here: https://github.com/lightningnetwork/lnd/blob/e8c5e7d5ce4b79117ef9ab497745fea3a1bffb36/invoices/invoiceregistry.go#L889

It is def a bit ugly that we have these logs, but perhaps we can keep them? wdyt @guggero ?

**Comment by aakselrod** [comment]
Should we put my last 2 committs in anothet PR as well? They're also unrelated to the original fix.

**Comment by bhandras** [comment]
> Should we put my last 2 committs in anothet PR as well? They're also unrelated to the original fix.

PR opened here: https://github.com/lightningnetwork/lnd/pull/9050

**Comment by guggero** [comment]
Closing in favor of https://github.com/lightningnetwork/lnd/pull/9050.

**Pull request closed without merging**